libsignal-service/src/provisioning/manager.rs

```diff
@@ -235,48 +171,145 @@ impl<P: PushService> LinkingManager<P> {
-                    );
-
                     let provisioning_code = message.provisioning_code.ok_or(
                         ProvisioningError::InvalidData {
                             reason: "no provisioning confirmation code".into(),
                         },
                     )?;
 
-                    let device_id = provisioning_manager
-                        .confirm_device(
-                            &provisioning_code,
-                            ConfirmDeviceMessage {
-                                signaling_key: signaling_key.to_vec(),
-                                supports_sms: false,
-                                fetches_messages: true,
-                                registration_id,
-                                pni_registration_id,
-                                name: vec![],
-                            },
+                    let aci_key_pair =
+                        KeyPair::new(aci_public_key, aci_private_key);
+                    let pni_key_pair =
+                        KeyPair::new(pni_public_key, pni_private_key);
+
+                    let aci_pq_last_resort_pre_key =
+                        generate_last_resort_kyber_key(
+                            aci_store,
+                            &aci_key_pair,
+                        )
+                        .await?;
+                    let pni_pq_last_resort_pre_key =
+                        generate_last_resort_kyber_key(
+                            pni_store,
+                            &pni_key_pair,
                         )
                         .await?;
 
-                    tx.send(
-                        SecondaryDeviceProvisioning::NewDeviceRegistration {
-                            phone_number,
-                            device_id,
+                    let aci_signed_pre_key = generate_signed_pre_key(
+                        aci_store,
+                        csprng,
+                        &aci_key_pair,
+                    )
+                    .await?;
+                    let pni_signed_pre_key = generate_signed_pre_key(
+                        pni_store,
+                        csprng,
+                        &pni_key_pair,
+                    )
+                    .await?;
+
+                    #[derive(Debug, Serialize)]
+                    #[serde(rename_all = "camelCase")]
+                    struct LinkRequest {
+                        verification_code: String,
+                        account_attributes: AccountAttributes,
+                        aci_signed_pre_key: SignedPreKey,
+                        pni_signed_pre_key: SignedPreKey,
+                        aci_pq_last_resort_pre_key: KyberPreKeyEntity,
+                        pni_pq_last_resort_pre_key: KyberPreKeyEntity,
+                    }
+
+                    #[derive(Debug, Serialize)]
+                    #[serde(rename_all = "camelCase")]
+                    struct AccountAttributes {
```

Comment:
I created this struct just for testing. There are already account attributes and capabilities. `DeviceCapabilities` are missing the `pni` field though.

Reply:
However I did not test if it links with these structs successfully. Maybe, they have some hard deprecated fields or similar. So, definitely need to test in case you decide to replace them.

Reply:
Before we can cleanup, we need to check how we can fix/adapt registration as primary device, so I'd defer this to a later stage.